lib/react_client.dart

```diff
@@ -55,37 +34,53 @@ abstract class ReactComponentFactoryProxy implements Function {
   get type;
 
   /// Returns a new rendered component instance with the specified [props] and [children].
-  JsObject call(Map props, [dynamic children]);
+  ReactElement call(Map props, [dynamic children]);
 
   /// Used to implement a variadic version of [call], in which children may be specified as additional arguments.
   dynamic noSuchMethod(Invocation invocation);
 }
 
+/// Prepares [children] to be passed to ReactJS.
+///
+/// Currently only involves converting a top-level non-[List] [Iterable] to
+/// a non-growable [List]s, but this may be updated in the future to support
+/// advanced nesting and other kinds of children.
+dynamic jsifyChildren(dynamic children) {
+  if (children is Iterable && children is! List) {
+    return children.toList(growable: false);
+  } else {
+    return children;
+  }
+}
+
 /// Creates ReactJS [Component] instances for Dart components.
-class ReactDartComponentFactoryProxy extends ReactComponentFactoryProxy {
-  final JsFunction reactClass;
-  final JsFunction reactComponentFactory;
+class ReactDartComponentFactoryProxy<TComponent extends Component> extends ReactComponentFactoryProxy {
+  final ReactClass reactClass;
+  final Function reactComponentFactory;
+  final Map defaultProps;
 
-  ReactDartComponentFactoryProxy(JsFunction reactClass) :
+  ReactDartComponentFactoryProxy(ReactClass reactClass) :
       this.reactClass = reactClass,
-      this.reactComponentFactory = _React.callMethod('createFactory', [reactClass]);
+      this.reactComponentFactory = React.createFactory(reactClass),
+      this.defaultProps = reactClass.dartDefaultProps
+  {
+    assert(defaultProps != null);
+  }
 
-  JsFunction get type => reactClass;
+  ReactClass get type => reactClass;
 
-  JsObject call(Map props, [dynamic children]) {
-    // Convert Iterable children to JsArrays so that the JS can read them.
-    // Use JsArrays instead of Lists, because automatic List conversion results in
-    // react-id values being cluttered with ".$o:0:0:$_jsObject:" in dart2js-transpiled Dart.
-    if (children is Iterable) {
-      children = new JsArray.from(children);
+  ReactElement<TComponent> call(Map props, [dynamic children]) {
+    // Convert Iterable children to Lists so that the JS can read them,
+    // and so they don't get iterated twice when passed to the Dart component
+    // and to the JS component.
+    if (children is Iterable && children is! List) {
```

Comment:
Functionally? Yes. Semantically, not quite... 

The purpose of `jsifyChildren` is to prepare children to be passed to ReactJs, whereas the purpose of this block is to prepare children to be passed to the Dart component.

Just noticed that the comment is misleading, as well:
> Convert Iterable children to Lists so that the JS can read them

---

I think I'll rename `jsifyChildren` to something like `listifyChildren`/`iterateChildren`, removing the method's semantic purpose as preparing children for JS, and then reuse its return value in both places.